File: packages/swift-code-block/Sources/SwiftCodeBlock/Render/CodeBlock.swift
#if canImport(SwiftUI)
    import SwiftUI

    public struct CodeBlockTheme {
        public var colors: [String: Color]
        public var background: Color

        public init(colors: [String: Color], background: Color = Color(red: 0.12, green: 0.12, blue: 0.12)) {
            self.colors = colors
            self.background = background
        }

        public static let standard = CodeBlockTheme(
            colors: [
                "plain": Color(red: 0.83, green: 0.83, blue: 0.83),
                "comment": Color(red: 0.42, green: 0.60, blue: 0.33),
                "keyword": Color(red: 0.77, green: 0.53, blue: 0.80),
                "string": Color(red: 0.81, green: 0.57, blue: 0.47),
                "number": Color(red: 0.71, green: 0.81, blue: 0.66),
                "function": Color(red: 0.86, green: 0.86, blue: 0.67),
                "class-name": Color(red: 0.31, green: 0.79, blue: 0.69),
                "operator": Color(red: 0.83, green: 0.83, blue: 0.83),
                "punctuation": Color(red: 0.83, green: 0.83, blue: 0.83),
            ]
        )
    }

    public struct CodeBlock: View {
        private let code: String
        private let language: String
        private let fallbackLanguage: String
        private let theme: CodeBlockTheme

        @ScaledMetric(relativeTo: .body) private var scaledFontSize: CGFloat = 14
        @State private var renderedState: CodeRenderState = .pending

        public init(
            _ code: String,
            language: String = "plain",
            fallbackLanguage: String = "plain",
            fontSize: CGFloat = 14,
            theme: CodeBlockTheme = .standard
        ) {
            self.code = code
            self.language = language
            self.fallbackLanguage = fallbackLanguage
            self.theme = theme
            _scaledFontSize = ScaledMetric(wrappedValue: fontSize, relativeTo: .body)
        }

        public var body: some View {
            ScrollView([.horizontal, .vertical]) {
                content(for: renderedState)
                    .font(.system(size: scaledFontSize, design: .monospaced))
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(12)
            }
            .background(theme.background)
            .clipShape(RoundedRectangle(cornerRadius: 8, style: .continuous))
            .accessibilityLabel(Text("\(language) code block"))
            .task(id: taskID) {
                for await state in CodeRenderer.render(
                    code: code,
                    options: CodeRenderOptions(
                        language: language,
                        fallbackLanguage: fallbackLanguage,
                        fontSize: Double(scaledFontSize),
                        scale: 1
                    )
                ) {
                    renderedState = state
                }
            }
        }

        private var taskID: String {
            [
                code,
                language,
                fallbackLanguage,
                "\(scaledFontSize)",
            ].joined(separator: "\u{1F}")
        }

        @ViewBuilder
        private func content(for state: CodeRenderState) -> some View {
            switch state {
            case .pending:
                Color.clear.frame(minWidth: 1, minHeight: max(scaledFontSize, 1))
            case .succeeded(let rendered):
                highlightedText(rendered.tokens)
                    .textSelection(.enabled)
            case .failed(let fallback):
                Text(fallback.text)
                    .foregroundColor(theme.colors["plain"] ?? .primary)
                    .textSelection(.enabled)
            }
        }

        private func highlightedText(_ tokens: [CodeTokenRun]) -> Text {
            tokens.reduce(Text("")) { partial, token in
                partial + Text(token.text).foregroundColor(color(for: token))
            }
        }

        private func color(for token: CodeTokenRun) -> Color {
            for type in token.types.reversed() {
                if let color = theme.colors[type] {
                    return color
                }
            }
            return theme.colors["plain"] ?? .primary
        }
    }
#endif

File: packages/compose-code-block/src/main/kotlin/io/github/dongyuzhao/composecodeblock/CodeBlock.kt
package io.github.dongyuzhao.composecodeblock

import androidx.compose.foundation.background
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

@Immutable
data class CodeBlockTheme(
    val colors: Map<String, Color>,
    val background: Color = Color(0xFF1E1E1E)
) {
    fun colorFor(token: CodeTokenRun, fallback: Color): Color {
        for (type in token.types.asReversed()) {
            colors[type]?.let { return it }
        }
        return colors["plain"] ?: fallback
    }

    companion object {
        val Standard = CodeBlockTheme(
            colors = mapOf(
                "plain" to Color(0xFFD4D4D4),
                "comment" to Color(0xFF6A9955),
                "keyword" to Color(0xFFC586C0),
                "string" to Color(0xFFCE9178),
                "number" to Color(0xFFB5CEA8),
                "function" to Color(0xFFDCDCAA),
                "class-name" to Color(0xFF4EC9B0),
                "operator" to Color(0xFFD4D4D4),
                "punctuation" to Color(0xFFD4D4D4)
            )
        )
    }
}

@Composable
fun CodeBlock(
    code: String,
    modifier: Modifier = Modifier,
    options: CodeRenderOptions = CodeRenderOptions(),
    theme: CodeBlockTheme = CodeBlockTheme.Standard,
    color: Color = LocalContentColor.current
) {
    val context = LocalContext.current
    val runtime = remember(context) { WebViewPrismRuntime(context) }
    val bridge = remember(runtime) { PrismBridge(runtime) }
    val renderer = remember(bridge) { CodeRenderer(bridge) }
    var renderState by remember { mutableStateOf<CodeRenderState>(CodeRenderState.Pending) }

    DisposableEffect(runtime) {
        onDispose { runtime.close() }
    }

    LaunchedEffect(code, options) {
        renderer.render(code, options).collect { state ->
            renderState = state
        }
    }

    Box(
        modifier = modifier
            .clip(RoundedCornerShape(8.dp))
            .background(theme.background)
            .horizontalScroll(rememberScrollState())
            .padding(12.dp)
    ) {
        when (val state = renderState) {
            CodeRenderState.Pending -> {
                Box(
                    modifier = Modifier
                        .testTag("code-block-pending")
                        .defaultMinSize(minWidth = 1.dp, minHeight = 1.dp)
                )
            }
            is CodeRenderState.Succeeded -> {
                Text(
                    text = annotatedCode(state.rendered.tokens, theme, color),
                    modifier = Modifier.testTag("code-block-success"),
                    style = TextStyle(
                        fontFamily = FontFamily.Monospace,
                        fontSize = (finitePositive(options.fontSizeSp, 14f) *
                            finitePositive(options.scale, 1f)).sp
                    )
                )
            }
            is CodeRenderState.Failed -> {
                Text(
                    text = state.fallback.text,
                    modifier = Modifier.testTag("code-block-failed"),
                    color = color,
                    style = TextStyle(
                        fontFamily = FontFamily.Monospace,
                        fontSize = finitePositive(options.fontSizeSp, 14f).sp
                    )
                )
            }
        }
    }
}

private fun annotatedCode(
    tokens: List<CodeTokenRun>,
    theme: CodeBlockTheme,
    fallback: Color
) = buildAnnotatedString {
    for (token in tokens) {
        withStyle(SpanStyle(color = theme.colorFor(token, fallback))) {
            append(token.text)
        }
    }
}

private fun finitePositive(value: Float, fallback: Float): Float {
    return if (value.isFinite() && value > 0f) value else fallback
}
